File: clients/python/src/ate_env/env.py
# Copyright 2026 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Handles to a single environment and to processes running inside it."""

from __future__ import annotations

from collections.abc import AsyncIterable, AsyncIterator, Iterable, Mapping, Sequence
from datetime import timedelta
from typing import TYPE_CHECKING

import grpc

from ._gen.ateenv.v1alpha import guest_pb2
from .errors import ProcessExitedError, map_rpc_error
from .types import (
    EnvironmentInfo,
    ProcessInfo,
    ProcessOutput,
    ShellResult,
    Signal,
    _process_info_from_pb,
    _process_output_from_pb,
)

if TYPE_CHECKING:
    from .client import Client

__all__ = ["Env", "Process"]

# Matches the server's file chunk size (guest/filesystem).
_CHUNK_SIZE = 64 * 1024

# Offsets past any spool: follow the stream without receiving output.
_SKIP_ALL = 2**63 - 1

_Bytes = bytes | bytearray | memoryview
_ByteSource = _Bytes | Iterable[bytes] | AsyncIterable[bytes]


class Env:
    """A handle to a single environment.

    Constructed by Client.create() or Client.env(). Guest operations
    (processes and files) automatically carry the x-env-id/x-env-atespace
    routing metadata the API server requires.
    """

    def __init__(self, client: Client, id: str, atespace: str):
        self._client = client
        self._id = id
        self._atespace = atespace

    @property
    def id(self) -> str:
        """The environment's identifier."""
        return self._id

    @property
    def atespace(self) -> str:
        """The environment's atespace."""
        return self._atespace

    def _metadata(self) -> tuple[tuple[str, str], ...]:
        return (("x-env-id", self._id), ("x-env-atespace", self._atespace))

    async def info(self) -> EnvironmentInfo:
        """Retrieve the environment's current status and configuration."""
        return await self._client.get(self._id, atespace=self._atespace)

    async def suspend(self) -> None:
        """Checkpoint and stop the environment."""
        await self._client.suspend(self._id, atespace=self._atespace)

    async def delete(self) -> None:
        """Remove the environment permanently."""
        await self._client.delete(self._id, atespace=self._atespace)

    # --- processes ---

    async def start_process(
        self,
        command: Sequence[str],
        *,
        cwd: str = "",
        env: Mapping[str, str] | None = None,
        stdin: bool = False,
        timeout: float | timedelta | None = None,
    ) -> Process:
        """Launch a process in the background and return a handle to it.

        stdin=True opens a pipe for standard input, fed with
        Process.write_input(); otherwise the process reads EOF immediately.
        timeout (seconds or timedelta) kills the process with SIGKILL when it
        elapses; None applies the guest default.
        """
        req = guest_pb2.StartProcessRequest(
            command=list(command), cwd=cwd, env=dict(env or {}), stdin=stdin
        )
        if timeout is not None:
            req.timeout.FromTimedelta(
                timeout if isinstance(timeout, timedelta) else timedelta(seconds=timeout)
            )
        try:
            proc = await self._client._processes.StartProcess(req, metadata=self._metadata())
        except grpc.RpcError as e:
            raise map_rpc_error(e) from e
        return Process(self, proc.process_id)

    def process(self, process_id: str) -> Process:
        """Return a handle to an existing process without checking that it exists."""
        return Process(self, process_id)

    async def shell(
        self,
        command_line: str,
        *,
        stdin: bytes | str | None = None,
        cwd: str = "",
        env: Mapping[str, str] | None = None,
        timeout: float | timedelta | None = None,
    ) -> ShellResult:
        """Run `sh -c command_line`, feed it stdin, and capture output and exit status.

        Output is buffered fully in memory; for incremental output or
        interactive input use start_process().
        """
        proc = await self.start_process(
            ["sh", "-c", command_line], cwd=cwd, env=env, stdin=stdin is not None, timeout=timeout
        )
        if stdin is not None:
            await proc.write_input(stdin.encode() if isinstance(stdin, str) else stdin, close=True)

        stdout = bytearray()
        stderr = bytearray()
        exit_info: ProcessInfo | None = None
        async for out in proc.output(follow=True):
            if out.stdout is not None:
                stdout.extend(out.stdout)
            elif out.stderr is not None:
                stderr.extend(out.stderr)
            elif out.exit is not None:
                exit_info = out.exit
        if exit_info is None:
            raise RuntimeError("ate_env: output stream ended before the process exited")
        return ShellResult(
            stdout=stdout.decode("utf-8", errors="replace"),
            stderr=stderr.decode("utf-8", errors="replace"),
            exit_code=exit_info.exit_code,
        )

    # --- files ---

    async def read_file(self, path: str) -> AsyncIterator[bytes]:
        """Stream the contents of a file in chunks.

        Errors (e.g. NotFoundError) surface on the first iteration.
        """
        req = guest_pb2.ReadFileRequest(path=path)
        call = self._client._filesystem.ReadFile(req, metadata=self._metadata())
        try:
            async for msg in call:
                if msg.chunk:
                    yield msg.chunk
        except grpc.RpcError as e:
            raise map_rpc_error(e) from e
        finally:
            call.cancel()

    async def read_file_bytes(self, path: str) -> bytes:
        """Read a whole file into memory."""
        return b"".join([chunk async for chunk in self.read_file(path)])

    async def write_file(
        self,
        path: str,
        data: _ByteSource,
        *,
        mode: int = 0o644,
        seek_offset: int = 0,
    ) -> int:
        """Write data to a file; returns the number of bytes written.

        data may be bytes-like (sent in 64 KiB chunks) or a sync/async
        iterable of bytes chunks. With seek_offset=0 (the default) the file
        is replaced. A positive seek_offset keeps existing content and
        starts writing at that byte, extending the file with zeros if it is
        shorter. mode applies when the file is created.
        """
        _validate_byte_source(data, "write_file")
        if seek_offset < 0:
            raise ValueError(f"ate_env: seek_offset must be >= 0, got {seek_offset}")
        try:
            resp = await self._client._filesystem.WriteFile(
                self._write_requests(path, data, mode, seek_offset), metadata=self._metadata()
            )
        except grpc.RpcError as e:
            raise map_rpc_error(e) from e
        return resp.bytes_written

    async def _write_requests(
        self, path: str, data: _ByteSource, mode: int, seek_offset: int
    ) -> AsyncIterator[guest_pb2.WriteFileRequest]:
        chunks = _chunk_stream(data)
        # The first message carries path, mode and seek_offset, even for
        # empty input (required to create empty files).
        first = b""
        async for chunk in chunks:
            first = chunk
            break
        yield guest_pb2.WriteFileRequest(
            path=path, mode=mode & 0o777, chunk=first, seek_offset=seek_offset
        )
        async for chunk in chunks:
            if chunk:
                yield guest_pb2.WriteFileRequest(chunk=chunk)


class Process:
    """A handle to a process inside an environment.

    Constructed by Env.start_process() or Env.process().
    """

    def __init__(self, env: Env, process_id: str):
        self._env = env
        self._id = process_id

    @property
    def id(self) -> str:
        """The guest-assigned process identifier."""
        return self._id

    @property
    def env(self) -> Env:
        """The environment the process runs in."""
        return self._env

    def __repr__(self) -> str:
        return f"Process({self._id!r})"

    @property
    def _stub(self):
        return self._env._client._processes

    async def info(self) -> ProcessInfo:
        """Retrieve the current state of the process."""
        req = guest_pb2.GetProcessRequest(process_id=self._id)
        try:
            proc = await self._stub.GetProcess(req, metadata=self._env._metadata())
        except grpc.RpcError as e:
            raise map_rpc_error(e) from e
        return _process_info_from_pb(proc)

    async def output(
        self,
        *,
        follow: bool = False,
        stdout_offset: int = 0,
        stderr_offset: int = 0,
    ) -> AsyncIterator[ProcessOutput]:
        """Stream stdout and stderr; the final message carries the exit state.

        With follow=True the stream stays open until the process exits, so on
        a long-lived process it blocks until then; consume it under a timeout
        or cancel the task. Without follow, the output spooled so far is
        returned and the exit message is present only if the process has
        already exited. Breaking out of the loop cancels the underlying RPC.
        """
        req = guest_pb2.StreamProcessOutputRequest(
            process_id=self._id,
            stdout_offset=stdout_offset,
            stderr_offset=stderr_offset,
            follow=follow,
        )
        call = self._stub.StreamProcessOutput(req, metadata=self._env._metadata())
        try:
            async for msg in call:
                yield _process_output_from_pb(msg)
        except grpc.RpcError as e:
            raise map_rpc_error(e) from e
        finally:
            call.cancel()

    async def wait(self) -> ProcessInfo:
        """Block until the process exits and return its final state.

        Follows the output stream with offsets past the end of the spool, so
        no output is transferred.
        """
        async for out in self.output(
            follow=True, stdout_offset=_SKIP_ALL, stderr_offset=_SKIP_ALL
        ):
            if out.exit is not None:
                return out.exit
        raise RuntimeError("ate_env: output stream ended before the process exited")

    async def write_input(self, data: _ByteSource, *, close: bool = False) -> int:
        """Write to the process's stdin; returns the number of bytes written.

        The process must have been started with stdin=True. stdin stays open
        across calls until one sets close=True (EOF). data may be bytes-like
        or a sync/async iterable of bytes chunks.
        """
        _validate_byte_source(data, "write_input")
        try:
            resp = await self._stub.WriteProcessInput(
                self._input_requests(data, close), metadata=self._env._metadata()
            )
        except grpc.RpcError as e:
            raise map_rpc_error(e) from e
        return resp.bytes_written

    async def close_input(self) -> None:
        """Close the process's stdin, delivering EOF."""
        await self.write_input(b"", close=True)

    async def _input_requests(
        self, data: _ByteSource, close: bool
    ) -> AsyncIterator[guest_pb2.WriteProcessInputRequest]:
        first = True
        async for chunk in _chunk_stream(data):
            if not chunk:
                continue
            req = guest_pb2.WriteProcessInputRequest(data=chunk)
            if first:
                req.process_id = self._id
                first = False
            yield req
        if first or close:
            # Either nothing was sent (the first message must name the
            # process) or EOF was requested; both need a final message.
            yield guest_pb2.WriteProcessInputRequest(
                process_id=self._id if first else "", close=close
            )

    async def signal(self, sig: Signal) -> None:
        """Deliver a signal to the process and its process group."""
        req = guest_pb2.SignalProcessRequest(process_id=self._id, signal=int(sig))
        try:
            await self._stub.SignalProcess(req, metadata=self._env._metadata())
        except grpc.RpcError as e:
            raise map_rpc_error(e) from e

    async def kill(self) -> ProcessInfo:
        """Send SIGKILL and wait for the process to exit.

        Killing a process that has already exited is not an error.
        """
        try:
            await self.signal(Signal.KILL)
        except ProcessExitedError:
            pass
        return await self.wait()


def _validate_byte_source(data: object, what: str) -> None:
    # Validate eagerly: errors raised inside a request generator are
    # swallowed by grpc.aio and surface as an opaque CancelledError.
    if isinstance(data, str):
        raise TypeError(f"ate_env: {what} expects bytes, not str; encode it first")
    if not isinstance(data, (bytes, bytearray, memoryview, Iterable, AsyncIterable)):
        raise TypeError(f"ate_env: unsupported {what} data type: {type(data)!r}")


async def _chunk_stream(data: _ByteSource) -> AsyncIterator[bytes]:
    if isinstance(data, (bytes, bytearray, memoryview)):
        buf = bytes(data)
        for i in range(0, len(buf), _CHUNK_SIZE):
            yield buf[i : i + _CHUNK_SIZE]
    elif isinstance(data, AsyncIterable):
        async for chunk in data:
            yield bytes(chunk)
    elif isinstance(data, Iterable):
        for chunk in data:
            yield bytes(chunk)
    else:
        raise TypeError(f"ate_env: unsupported data type: {type(data)!r}")

File: clients/go/env.go
// Copyright 2026 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package env

import (
	"bytes"
	"context"
	"errors"
	"fmt"
	"io"
	"io/fs"

	ateenvv1alpha "github.com/agent-substrate/env/proto/ateenv/v1alpha"
	"google.golang.org/grpc/metadata"
	"google.golang.org/protobuf/types/known/durationpb"
)

// Env is a handle to a single environment.
type Env struct {
	id       string
	atespace string
	client   *Client
}

// ID returns the environment's identifier.
func (e *Env) ID() string { return e.id }

func (e *Env) withEnv(ctx context.Context) context.Context {
	return metadata.AppendToOutgoingContext(ctx, "x-env-id", e.id, "x-env-atespace", e.atespace)
}

// Suspend checkpoints and stops the environment using gRPC.
func (e *Env) Suspend(ctx context.Context) error {
	return e.client.Suspend(ctx, e.atespace, e.id)
}

// Delete removes the environment permanently using gRPC.
func (e *Env) Delete(ctx context.Context) error {
	return e.client.Delete(ctx, e.atespace, e.id)
}

// Shell runs a shell command line inside the environment and captures its
// output. It is shorthand for Run with only Command set.
func (e *Env) Shell(ctx context.Context, commandLine string) (*ShellResponse, error) {
	return e.Run(ctx, ShellRequest{Command: commandLine})
}

// Run executes `sh -c req.Command` inside the environment, feeds it
// req.Stdin, and returns its buffered output and exit status once it exits.
// For incremental output or interactive input use StartProcess.
func (e *Env) Run(ctx context.Context, req ShellRequest) (*ShellResponse, error) {
	start := &ateenvv1alpha.StartProcessRequest{
		Command: []string{"sh", "-c", req.Command},
		Cwd:     req.Cwd,
		Env:     req.Env,
		Stdin:   req.Stdin != nil,
	}
	if req.Timeout > 0 {
		start.Timeout = durationpb.New(req.Timeout)
	}
	proc, err := e.StartProcess(ctx, start)
	if err != nil {
		return nil, err
	}

	if req.Stdin != nil {
		w, err := proc.Stdin(ctx)
		if err != nil {
			return nil, err
		}
		if _, err := w.Write(req.Stdin); err != nil {
			return nil, err
		}
		if err := w.Close(); err != nil {
			return nil, err
		}
	}

	stream, err := proc.Output(ctx, &ateenvv1alpha.StreamProcessOutputRequest{Follow: true})
	if err != nil {
		return nil, err
	}
	var stdoutBuf, stderrBuf bytes.Buffer
	var exit *ateenvv1alpha.Process
	for {
		msg, err := stream.Recv()
		if errors.Is(err, io.EOF) {
			break
		}
		if err != nil {
			return nil, fromGRPCError(err)
		}
		switch out := msg.GetOutput().(type) {
		case *ateenvv1alpha.ProcessOutput_Stdout:
			stdoutBuf.Write(out.Stdout)
		case *ateenvv1alpha.ProcessOutput_Stderr:
			stderrBuf.Write(out.Stderr)
		case *ateenvv1alpha.ProcessOutput_Exit:
			exit = out.Exit
		}
	}
	if exit == nil {
		return nil, errors.New("env: output stream ended before the process exited")
	}

	return &ShellResponse{
		Stdout:   stdoutBuf.String(),
		Stderr:   stderrBuf.String(),
		ExitCode: int(exit.GetExitCode()),
	}, nil
}

// ReadFile streams the contents of the file at path inside the environment using FileSystemService.
// The caller must close the returned reader.
func (e *Env) ReadFile(ctx context.Context, p string) (io.ReadCloser, error) {
	ctx = e.withEnv(ctx)
	stream, err := e.client.filesystem.ReadFile(ctx, &ateenvv1alpha.ReadFileRequest{
		Path: p,
	})
	if err != nil {
		return nil, fromGRPCError(err)
	}

	// Read the first chunk eagerly to check for immediate errors (e.g. NotFound, PermissionDenied)
	firstChunk, err := stream.Recv()
	if errors.Is(err, io.EOF) {
		// Empty file
		return io.NopCloser(bytes.NewReader(nil)), nil
	}
	if err != nil {
		return nil, fromGRPCError(err)
	}

	pr, pw := io.Pipe()
	go func() {
		if len(firstChunk.GetChunk()) > 0 {
			if _, writeErr := pw.Write(firstChunk.GetChunk()); writeErr != nil {
				return
			}
		}
		for {
			chunk, err := stream.Recv()
			if errors.Is(err, io.EOF) {
				_ = pw.Close()
				return
			}
			if err != nil {
				_ = pw.CloseWithError(fromGRPCError(err))
				return
			}
			if len(chunk.GetChunk()) > 0 {
				if _, writeErr := pw.Write(chunk.GetChunk()); writeErr != nil {
					return
				}
			}
		}
	}()

	return pr, nil
}

// WriteFile replaces the file at path inside the environment with the
// contents of r, creating it with the given permissions if needed.
func (e *Env) WriteFile(ctx context.Context, p string, r io.Reader, mode fs.FileMode) error {
	return e.writeFile(ctx, p, r, mode, 0)
}

// WriteFileAt writes the contents of r into the file at path starting at
// byte offset, keeping existing content outside the written range. The file
// is created with the given permissions if it does not exist, and extended
// with zero bytes if offset is past its end. An offset of zero replaces the
// file, like WriteFile.
func (e *Env) WriteFileAt(ctx context.Context, p string, offset int64, r io.Reader, mode fs.FileMode) error {
	if offset < 0 {
		return fmt.Errorf("env: negative offset %d for %q", offset, p)
	}
	return e.writeFile(ctx, p, r, mode, offset)
}

func (e *Env) writeFile(ctx context.Context, p string, r io.Reader, mode fs.FileMode, seekOffset int64) error {
	ctx = e.withEnv(ctx)
	stream, err := e.client.filesystem.WriteFile(ctx)
	if err != nil {
		return fromGRPCError(err)
	}

	buf := make([]byte, 64*1024)
	n, readErr := r.Read(buf)
	if readErr != nil && !errors.Is(readErr, io.EOF) {
		return fmt.Errorf("env: reading data for %q: %w", p, readErr)
	}

	firstReq := &ateenvv1alpha.WriteFileRequest{
		Path:       p,
		Mode:       uint32(mode.Perm()),
		Chunk:      buf[:n],
		SeekOffset: seekOffset,
	}
	if err := stream.Send(firstReq); err != nil {
		return fromGRPCError(err)
	}

	if !errors.Is(readErr, io.EOF) {
		for {
			n, err := r.Read(buf)
			if n > 0 {
				if sendErr := stream.Send(&ateenvv1alpha.WriteFileRequest{Chunk: buf[:n]}); sendErr != nil {
					return fromGRPCError(sendErr)
				}
			}
			if errors.Is(err, io.EOF) {
				break
			}
			if err != nil {
				return fmt.Errorf("env: reading data for %q: %w", p, err)
			}
		}
	}

	if _, err := stream.CloseAndRecv(); err != nil {
		return fromGRPCError(err)
	}
	return nil
}
